### PR TITLE
fix: merge artifact arguments from workflow template. Fixes #4296

### LINKED ISSUE #4296: Cannot submit workflow from workflow template with input artifact argument
## Summary 
I've converted https://github.com/argoproj/argo/blob/master/examples/arguments-artifacts.yaml to workflow template:
```
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: arguments-artifacts
spec:
  entrypoint: kubectl-input-artifact
  arguments:
    artifacts:
    - name: kubectl
      http:
        url: https://storage.googleapis.com/kubernetes-release/release/v1.8.0/bin/linux/amd64/kubectl

  templates:
  - name: kubectl-input-artifact
    inputs:
      artifacts:
      - name: kubectl
        path: /usr/local/bin/kubectl
        mode: 0755
    container:
      image: debian:9.4
      command: [sh, -c]
      args: ["kubectl version"]
```
 But when I try and submit any workflow using this template, submission fails
```
❯ argo  submit --from workflowtemplate/arguments-artifacts
2020/10/15 16:46:00 Failed to submit workflow: templates.kubectl-input-artifact inputs.artifacts.kubectl was not supplied
```

## Server/CLI version
2.11.3

### PR DESCRIPTION
**Pull request by istepura**
Fixes #4296 

Checklist:
* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not.

### CODE REVIEW

File: util/util.go [hunk 1]
```diff
@@ -83,6 +83,22 @@ func MergeParameters(params ...[]wfv1.Parameter) []wfv1.Parameter {
...
 }
 
+// MergeArtifacts merges artifact argument slices
+// Merge the slices based on arguments order (first is high priority).
+func MergeArtifacts(artifactSlices ...[]wfv1.Artifact) []wfv1.Artifact {
+	var result []wfv1.Artifact
+	alreadyMerged := make(map[string]bool)
+	for _, artifacts := range artifactSlices {
+		for _, item := range artifacts {
+			if _, found := alreadyMerged[item.Name]; !found {
```
Comment (alexec) [inline util/util.go:93]: minor: with a bool map, missing entries return false, so you can do this

```golang
if !alreadyMerged[item.Name] {
```

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -3137,6 +3137,10 @@ func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wf
 		executionParameters.Parameters = util.MergeParameters(executionParameters.Parameters, woc.execWf.Spec.Arguments.Parameters)
 	}
 
+	if len(woc.execWf.Spec.Arguments.Artifacts) > 0 {
+		executionParameters.Artifacts = util.MergeArtifacts(executionParameters.Artifacts, woc.execWf.Spec.Arguments.Artifacts)
```
Comment (alexec) [inline workflow/controller/operator.go:3141]: minor: don't think you the guard condition here

File: workflow/controller/operator_workflow_template_ref_test.go [hunk 3]
```diff
@@ -277,3 +277,76 @@ func TestWorkflowTemplateRefParamMerge(t *testing.T) {
...
+  workflowTemplateRef:
+    name: artifact-test-1
+`
+
+func TestWorkflowTemplateRefGetArtifactsFromTemplate(t *testing.T) {
+	wf := unmarshalWF(wfWithTemplateWithArtifact)
+	wftmpl := unmarshalWFTmpl(wftWithArtifact)
+
+	t.Run("CheckArtifactArgumentFromWF", func(t *testing.T) {
+		cancel, controller := newController(wf, wftmpl)
```
Comment (alexec) [inline workflow/controller/operator_workflow_template_ref_test.go:340]: nice

File: workflow/controller/operator_workflow_template_ref_test.go [hunk 4]
```diff
@@ -277,3 +277,76 @@ func TestWorkflowTemplateRefParamMerge(t *testing.T) {
...
+		defer cancel()
+		woc := newWorkflowOperationCtx(wf, controller)
+		_, execArgs, err := woc.loadExecutionSpec()
+		assert.NoError(t, err)
+		assert.Equal(t, wf.Spec.Arguments.Artifacts, woc.wf.Spec.Arguments.Artifacts)
+		assert.Equal(t, 3, len(execArgs.Artifacts))
```
Comment (alexec) [inline workflow/controller/operator_workflow_template_ref_test.go:346]: minor: assert.Len

File: util/util.go (hunk 1 above)
Comment (istepura) [reply util/util.go]: Fixed. Thanks

File: workflow/controller/operator.go (hunk 2 above)
Comment (istepura) [reply workflow/controller/operator.go]: Fixed

File: workflow/controller/operator_workflow_template_ref_test.go (hunk 4 above)
Comment (istepura) [reply workflow/controller/operator_workflow_template_ref_test.go]: Changed to `assert.Len`

**Review by alexec — APPROVED** [review]

**Pull request merged**